include/swift/AST/ClangTypeConverter.h

```diff
@@ -0,0 +1,112 @@
+//===-- ClangTypeConverter.h - Converting Swift types to C types-*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file defines utilities for translating Swift types to C types.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_AST_CLANG_TYPE_CONVERTER_H
+#define SWIFT_AST_CLANG_TYPE_CONVERTER_H
+
+#include "swift/AST/Type.h"
+#include "swift/AST/TypeVisitor.h"
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/Type.h"
+
+namespace swift {
+
+/// Compute C types corresponding to Swift AST types.
+class ClangTypeConverter :
+    public TypeVisitor<ClangTypeConverter, clang::QualType> {
+
+  using super = TypeVisitor<ClangTypeConverter, clang::QualType>;
+
+  llvm::DenseMap<Type, clang::QualType> Cache;
+
+  bool StdlibTypesAreCached = false;
+
+  ASTContext &Context;
+
+  clang::ASTContext &ClangASTContext;
+
+  const llvm::Triple Triple;
+
+  ClangTypeConverter(const ClangTypeConverter &) = delete;
+  ClangTypeConverter &operator=(const ClangTypeConverter &) = delete;
+
+public:
+
+  using ConstructorArgs =
+      std::tuple<ASTContext &, clang::ASTContext &, llvm::Triple>;
+
+  /// Static helper to avoid duplication at the constructor's call-site.
+  static ConstructorArgs getConstructorArgs(ASTContext &ctx);
```

Comment:
Why this abstraction instead of just taking an `ASTContext &` and then pulling the components out in the constructor?

Reply:
IIUC, reference data members (such as `clang::ASTContext &ClangASTContext`) need to be set in the initializer list in the constructor. This means that I'd have to write a lambda which returns the `clang::ASTContext &` and duplicate most of that logic for initializing the `llvm::Triple` field (either in the initializer list or in the body of the constructor).

Reply:
You can use previously-initialized fields in the initializers for later fields.